File: src/lib/garden.ts
"use server";

import { and, count, desc, eq, inArray, or, sql } from "drizzle-orm";
import { revalidatePath } from "next/cache";
import { db } from "@/db";
import { gardenInvitations, gardenMembers, user } from "@/db/schema";
import { GARDEN_CAPACITY, type GardenPersonStatus } from "@/lib/garden-types";
import type { GardenInvitationView } from "@/lib/garden-types";
import { getUserProgress } from "@/lib/progress";
import {
  publishGardenChanged,
  publishInviteIncoming,
  publishInviteRemoved,
} from "@/lib/realtime-publish";
import { requireSession } from "@/lib/session";

function newId() {
  return crypto.randomUUID();
}

const MAX_GUESTS = GARDEN_CAPACITY - 1;

function revalidateGarden() {
  revalidatePath("/garden");
  revalidatePath("/");
  revalidatePath("/habits");
}

function statusFromUser(row: {
  statusPreset: string | null;
  statusNote: string | null;
  statusUpdatedAt: Date | null;
}): GardenPersonStatus {
  return {
    statusPreset: row.statusPreset,
    statusNote: row.statusNote,
    statusUpdatedAt: row.statusUpdatedAt
      ? row.statusUpdatedAt.toISOString()
      : null,
  };
}

async function guestCountFor(ownerId: string) {
  const [{ value }] = await db
    .select({ value: count() })
    .from(gardenMembers)
    .where(eq(gardenMembers.ownerId, ownerId));
  return value;
}

async function ensureMembership(ownerId: string, memberId: string) {
  const existing = await db.query.gardenMembers.findFirst({
    where: and(
      eq(gardenMembers.ownerId, ownerId),
      eq(gardenMembers.memberId, memberId),
    ),
  });
  if (existing) return;
  await db.insert(gardenMembers).values({
    id: newId(),
    ownerId,
    memberId,
  });
}

export async function getGardenDashboard() {
  const session = await requireSession();
  const ownerId = session.user.id;

  const selfRow = await db.query.user.findFirst({
    where: eq(user.id, ownerId),
    columns: {
      statusPreset: true,
      statusNote: true,
      statusUpdatedAt: true,
    },
  });

  const selfProgress = await getUserProgress(ownerId);
  const self = {
    id: `self-${ownerId}`,
    memberId: ownerId,
    name: session.user.name,
    email: session.user.email,
    image: session.user.image ?? null,
    isSelf: true,
    ...selfProgress,
    ...statusFromUser({
      statusPreset: selfRow?.statusPreset ?? null,
      statusNote: selfRow?.statusNote ?? null,
      statusUpdatedAt: selfRow?.statusUpdatedAt ?? null,
    }),
  };

  const rows = await db
    .select({
      id: gardenMembers.id,
      memberId: gardenMembers.memberId,
      createdAt: gardenMembers.createdAt,
      name: user.name,
      email: user.email,
      image: user.image,
      statusPreset: user.statusPreset,
      statusNote: user.statusNote,
      statusUpdatedAt: user.statusUpdatedAt,
    })
    .from(gardenMembers)
    .innerJoin(user, eq(gardenMembers.memberId, user.id))
    .where(eq(gardenMembers.ownerId, ownerId))
    .orderBy(desc(gardenMembers.createdAt))
    .limit(MAX_GUESTS);

  const guests = await Promise.all(
    rows.map(async (row) => {
      const progress = await getUserProgress(row.memberId);
      return {
        id: row.id,
        memberId: row.memberId,
        name: row.name,
        email: row.email,
        image: row.image,
        isSelf: false,
        ...progress,
        ...statusFromUser(row),
      };
    }),
  );

  const [incoming, outgoing] = await Promise.all([
    getIncomingInvitations(),
    getOutgoingInvitations(),
  ]);

  return {
    people: [self, ...guests],
    guestCount: guests.length,
    maxGuests: MAX_GUESTS,
    capacity: GARDEN_CAPACITY,
    incomingInvites: incoming,
    outgoingInvites: outgoing,
  };
}

export async function getGardenStatuses(): Promise<
  Array<{ memberId: string } & GardenPersonStatus>
> {
  const session = await requireSession();
  const ownerId = session.user.id;

  const memberships = await db
    .select({ memberId: gardenMembers.memberId })
    .from(gardenMembers)
    .where(eq(gardenMembers.ownerId, ownerId))
    .limit(MAX_GUESTS);

  const ids = [ownerId, ...memberships.map((m) => m.memberId)];

  const rows = await db
    .select({
      id: user.id,
      statusPreset: user.statusPreset,
      statusNote: user.statusNote,
      statusUpdatedAt: user.statusUpdatedAt,
    })
    .from(user)
    .where(inArray(user.id, ids));

  return rows.map((row) => ({
    memberId: row.id,
    ...statusFromUser(row),
  }));
}

export async function getIncomingInvitations(): Promise<GardenInvitationView[]> {
  const session = await requireSession();
  const rows = await db
    .select({
      id: gardenInvitations.id,
      fromUserId: gardenInvitations.fromUserId,
      toUserId: gardenInvitations.toUserId,
      createdAt: gardenInvitations.createdAt,
      fromName: user.name,
      fromEmail: user.email,
    })
    .from(gardenInvitations)
    .innerJoin(user, eq(gardenInvitations.fromUserId, user.id))
    .where(
      and(
        eq(gardenInvitations.toUserId, session.user.id),
        eq(gardenInvitations.status, "pending"),
      ),
    )
    .orderBy(desc(gardenInvitations.createdAt));

  return rows.map((row) => ({
    id: row.id,
    fromUserId: row.fromUserId,
    toUserId: row.toUserId,
    fromName: row.fromName,
    fromEmail: row.fromEmail,
    toName: session.user.name,
    toEmail: session.user.email,
    createdAt: row.createdAt.toISOString(),
    direction: "incoming" as const,
  }));
}

export async function getOutgoingInvitations(): Promise<GardenInvitationView[]> {
  const session = await requireSession();
  const rows = await db
    .select({
      id: gardenInvitations.id,
      fromUserId: gardenInvitations.fromUserId,
      toUserId: gardenInvitations.toUserId,
      createdAt: gardenInvitations.createdAt,
      toName: user.name,
      toEmail: user.email,
    })
    .from(gardenInvitations)
    .innerJoin(user, eq(gardenInvitations.toUserId, user.id))
    .where(
      and(
        eq(gardenInvitations.fromUserId, session.user.id),
        eq(gardenInvitations.status, "pending"),
      ),
    )
    .orderBy(desc(gardenInvitations.createdAt));

  return rows.map((row) => ({
    id: row.id,
    fromUserId: row.fromUserId,
    toUserId: row.toUserId,
    fromName: session.user.name,
    fromEmail: session.user.email,
    toName: row.toName,
    toEmail: row.toEmail,
    createdAt: row.createdAt.toISOString(),
    direction: "outgoing" as const,
  }));
}

export async function getPendingInviteCount() {
  const session = await requireSession();
  const [{ value }] = await db
    .select({ value: count() })
    .from(gardenInvitations)
    .where(
      and(
        eq(gardenInvitations.toUserId, session.user.id),
        eq(gardenInvitations.status, "pending"),
      ),
    );
  return value;
}

/** Send a garden invitation (does not add membership until accepted). */
export async function inviteGardenMember(email: string) {
  const session = await requireSession();
  const trimmed = email.trim().toLowerCase();

  if (!trimmed || !trimmed.includes("@")) {
    return { error: "Enter a valid email" };
  }

  if (trimmed === session.user.email.toLowerCase()) {
    return { error: "You are already growing in your own garden" };
  }

  const myGuests = await guestCountFor(session.user.id);
  if (myGuests >= MAX_GUESTS) {
    return {
      error: `Garden is full (${GARDEN_CAPACITY} people including you)`,
    };
  }

  const found = await db.query.user.findFirst({
    where: sql`lower(${user.email}) = ${trimmed}`,
  });

  if (!found) {
    return { error: "No Bloom account with that email yet" };
  }

  const alreadyMember = await db.query.gardenMembers.findFirst({
    where: and(
      eq(gardenMembers.ownerId, session.user.id),
      eq(gardenMembers.memberId, found.id),
    ),
  });

  if (alreadyMember) {
    return { error: "They are already in your garden" };
  }

  const existingInvite = await db.query.gardenInvitations.findFirst({
    where: and(
      eq(gardenInvitations.fromUserId, session.user.id),
      eq(gardenInvitations.toUserId, found.id),
    ),
  });

  if (existingInvite?.status === "pending") {
    return { error: "Invitation already sent" };
  }

  const reversePending = await db.query.gardenInvitations.findFirst({
    where: and(
      eq(gardenInvitations.fromUserId, found.id),
      eq(gardenInvitations.toUserId, session.user.id),
      eq(gardenInvitations.status, "pending"),
    ),
  });

  if (reversePending) {
    return { error: "They already invited you — check your invitations" };
  }

  const invitationId = existingInvite?.id ?? newId();

  if (existingInvite) {
    await db
      .update(gardenInvitations)
      .set({
        status: "pending",
        createdAt: new Date(),
        respondedAt: null,
      })
      .where(eq(gardenInvitations.id, existingInvite.id));
  } else {
    await db.insert(gardenInvitations).values({
      id: invitationId,
      fromUserId: session.user.id,
      toUserId: found.id,
      status: "pending",
    });
  }

  publishInviteIncoming(found.id, {
    id: invitationId,
    fromUserId: session.user.id,
    toUserId: found.id,
    fromName: session.user.name,
    fromEmail: session.user.email,
    toName: found.name,
    toEmail: found.email,
    createdAt: new Date().toISOString(),
    direction: "incoming",
  });

  revalidateGarden();
  return { success: true as const, invitedName: found.name };
}

/** @deprecated use inviteGardenMember */
export async function addGardenMember(email: string) {
  return inviteGardenMember(email);
}

export async function acceptGardenInvitation(invitationId: string) {
  const session = await requireSession();

  const invite = await db.query.gardenInvitations.findFirst({
    where: and(
      eq(gardenInvitations.id, invitationId),
      eq(gardenInvitations.toUserId, session.user.id),
      eq(gardenInvitations.status, "pending"),
    ),
  });

  if (!invite) {
    return { error: "Invitation not found" };
  }

  const fromId = invite.fromUserId;
  const toId = invite.toUserId;

  const [fromGuests, toGuests] = await Promise.all([
    guestCountFor(fromId),
    guestCountFor(toId),
  ]);

  if (fromGuests >= MAX_GUESTS) {
    return { error: "Their garden is full" };
  }
  if (toGuests >= MAX_GUESTS) {
    return {
      error: `Your garden is full (${GARDEN_CAPACITY} people including you)`,
    };
  }

  await ensureMembership(fromId, toId);
  await ensureMembership(toId, fromId);

  await db
    .update(gardenInvitations)
    .set({ status: "accepted", respondedAt: new Date() })
    .where(eq(gardenInvitations.id, invitationId));

  // Clear any reverse pending invite between the same pair
  await db
    .delete(gardenInvitations)
    .where(
      and(
        eq(gardenInvitations.status, "pending"),
        or(
          and(
            eq(gardenInvitations.fromUserId, fromId),
            eq(gardenInvitations.toUserId, toId),
          ),
          and(
            eq(gardenInvitations.fromUserId, toId),
            eq(gardenInvitations.toUserId, fromId),
          ),
        ),
      ),
    );

  publishInviteRemoved([fromId, toId], invitationId);
  publishGardenChanged([fromId, toId], "invite-accepted");

  revalidateGarden();
  return { success: true as const };
}

export async function declineGardenInvitation(invitationId: string) {
  const session = await requireSession();

  const invite = await db.query.gardenInvitations.findFirst({
    where: and(
      eq(gardenInvitations.id, invitationId),
      eq(gardenInvitations.toUserId, session.user.id),
      eq(gardenInvitations.status, "pending"),
    ),
  });

  if (!invite) {
    return { error: "Invitation not found" };
  }

  await db
    .update(gardenInvitations)
    .set({ status: "declined", respondedAt: new Date() })
    .where(eq(gardenInvitations.id, invitationId));

  publishInviteRemoved(
    [session.user.id, invite.fromUserId],
    invitationId,
  );
  publishGardenChanged([invite.fromUserId], "invite-declined");

  revalidateGarden();
  return { success: true as const };
}

export async function cancelGardenInvitation(invitationId: string) {
  const session = await requireSession();

  const invite = await db.query.gardenInvitations.findFirst({
    where: and(
      eq(gardenInvitations.id, invitationId),
      eq(gardenInvitations.fromUserId, session.user.id),
      eq(gardenInvitations.status, "pending"),
    ),
  });

  if (!invite) {
    return { error: "Invitation not found" };
  }

  await db
    .delete(gardenInvitations)
    .where(eq(gardenInvitations.id, invitationId));

  publishInviteRemoved([session.user.id, invite.toUserId], invitationId);

  revalidateGarden();
  return { success: true as const };
}

export async function removeGardenMember(memberRowId: string) {
  const session = await requireSession();

  const row = await db.query.gardenMembers.findFirst({
    where: and(
      eq(gardenMembers.id, memberRowId),
      eq(gardenMembers.ownerId, session.user.id),
    ),
  });

  if (!row) {
    return { error: "Member not found" };
  }

  // Mutual remove — drop both directions
  await db
    .delete(gardenMembers)
    .where(
      or(
        and(
          eq(gardenMembers.ownerId, session.user.id),
          eq(gardenMembers.memberId, row.memberId),
        ),
        and(
          eq(gardenMembers.ownerId, row.memberId),
          eq(gardenMembers.memberId, session.user.id),
        ),
      ),
    );

  publishGardenChanged(
    [session.user.id, row.memberId],
    "member-removed",
  );

  revalidateGarden();
  return { success: true as const };
}

File: apps/api/app/routers/garden.py
from datetime import date, datetime, timedelta
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import and_, func, or_
from sqlalchemy.orm import Session as DbSession

from app import schemas as S
from app.deps import get_current_user, get_db
from app.domain.activity_stats import focus_summary_for_day
from app.domain.consistency import (
    CALENDAR_WINDOW,
    CONSISTENCY_WINDOW,
    compute_consistency,
    compute_daily_scores,
    compute_streak,
)
from app.domain.schedule import get_period_key
from app.models import (
    GardenInvitation,
    GardenMember,
    Habit,
    HabitCompletion,
    User,
)
from app.realtime.hub import hub

GARDEN_CAPACITY = 5
MAX_GUESTS = GARDEN_CAPACITY - 1

router = APIRouter(prefix="/garden", tags=["garden"])


def _date_keys(days: int, end: date | None = None) -> list[str]:
    end = end or date.today()
    return [(end - timedelta(days=days - 1 - i)).isoformat() for i in range(days)]


def _user_progress(db: DbSession, user_id: str) -> dict:
    fetch_days = max(CONSISTENCY_WINDOW, CALENDAR_WINDOW)
    window_start = _date_keys(fetch_days)[0]
    week_keys = _date_keys(7)
    consistency_start = _date_keys(CONSISTENCY_WINDOW)[0]

    active = (
        db.query(Habit)
        .filter(Habit.user_id == user_id, Habit.archived_at.is_(None))
        .order_by(Habit.created_at.desc())
        .all()
    )

    habit_ids = [h.id for h in active]
    completions = (
        db.query(HabitCompletion)
        .filter(
            HabitCompletion.user_id == user_id,
            HabitCompletion.habit_id.in_(habit_ids) if habit_ids else False,
            HabitCompletion.completed_on >= date.fromisoformat(window_start),
        )
        .all()
    ) if habit_ids else []

    comp_dicts = [
        {"habit_id": c.habit_id, "completed_on": c.completed_on.isoformat(), "slot": c.slot}
        for c in completions
    ]
    score_habits = [
        {
            "id": h.id,
            "frequency": h.frequency,
            "times_per_period": h.times_per_period,
            "custom_every_days": h.custom_every_days,
            "created_at": h.created_at,
        }
        for h in active
    ]

    today = date.today().isoformat()

    habit_stats = []
    for h in active:
        pk = get_period_key(h.frequency, h.times_per_period, h.custom_every_days, date.today(), h.created_at)
        times = max(1, h.times_per_period)
        period_slots = [c for c in completions if c.habit_id == h.id and c.completed_on.isoformat() == pk]
        week_done = len([c for c in completions if c.habit_id == h.id and c.completed_on.isoformat() >= week_keys[0]])
        window_done = len([c for c in completions if c.habit_id == h.id and c.completed_on.isoformat() >= consistency_start])
        habit_stats.append(S.GardenHabitStat(
            id=h.id, name=h.name,
            completedToday=len(period_slots) >= times,
            completedSlots=len(period_slots),
            timesPerPeriod=times,
            weekDone=min(week_done, 7 * times),
            windowRate=min(1.0, window_done / (CONSISTENCY_WINDOW * times)) if times else 0,
        ))

    completed_today = sum(1 for s in habit_stats if s.completedToday)
    focus = focus_summary_for_day(db, user_id)
    return {
        "habitCount": len(habit_ids),
        "completedToday": completed_today,
        "consistency": compute_consistency(score_habits, comp_dicts),
        "streak": compute_streak(score_habits, comp_dicts),
        "days": [S.DayScore(**d) for d in compute_daily_scores(score_habits, comp_dicts, CALENDAR_WINDOW)],
        "habits": habit_stats,
        **focus,
    }


def _status_from(u: User) -> dict:
    return {
        "statusPreset": u.status_preset,
        "statusNote": u.status_note,
        "statusUpdatedAt": u.status_updated_at.isoformat() if u.status_updated_at else None,
    }


def _guest_count(db: DbSession, owner_id: str) -> int:
    return db.query(func.count()).select_from(GardenMember).filter(GardenMember.owner_id == owner_id).scalar() or 0


def _ensure_membership(db: DbSession, owner_id: str, member_id: str) -> None:
    exists = db.query(GardenMember).filter(
        GardenMember.owner_id == owner_id, GardenMember.member_id == member_id,
    ).first()
    if not exists:
        db.add(GardenMember(id=uuid4().hex, owner_id=owner_id, member_id=member_id))


@router.get("/dashboard", response_model=S.GardenDashboard)
def dashboard(db: DbSession = Depends(get_db), user: User = Depends(get_current_user)):
    self_progress = _user_progress(db, user.id)
    self_person = S.GardenPerson(
        id=f"self-{user.id}", memberId=user.id, name=user.name,
        email=user.email, image=user.image, isSelf=True,
        **self_progress, **_status_from(user),
    )

    rows = (
        db.query(GardenMember, User)
        .join(User, GardenMember.member_id == User.id)
        .filter(GardenMember.owner_id == user.id)
        .order_by(GardenMember.created_at.desc())
        .limit(MAX_GUESTS)
        .all()
    )

    guests = []
    for gm, member in rows:
        progress = _user_progress(db, member.id)
        guests.append(S.GardenPerson(
            id=gm.id, memberId=member.id, name=member.name,
            email=member.email, image=member.image, isSelf=False,
            **progress, **_status_from(member),
        ))

    incoming = _incoming(db, user)
    outgoing = _outgoing(db, user)

    return S.GardenDashboard(
        people=[self_person, *guests],
        guestCount=len(guests), maxGuests=MAX_GUESTS, capacity=GARDEN_CAPACITY,
        incomingInvites=incoming, outgoingInvites=outgoing,
    )


def _incoming(db: DbSession, user: User) -> list[S.GardenInvitationView]:
    rows = (
        db.query(GardenInvitation, User)
        .join(User, GardenInvitation.from_user_id == User.id)
        .filter(GardenInvitation.to_user_id == user.id, GardenInvitation.status == "pending")
        .order_by(GardenInvitation.created_at.desc())
        .all()
    )
    return [
        S.GardenInvitationView(
            id=inv.id, fromUserId=inv.from_user_id, toUserId=inv.to_user_id,
            fromName=u.name, fromEmail=u.email, toName=user.name, toEmail=user.email,
            createdAt=inv.created_at.isoformat(), direction="incoming",
        )
        for inv, u in rows
    ]


def _outgoing(db: DbSession, user: User) -> list[S.GardenInvitationView]:
    rows = (
        db.query(GardenInvitation, User)
        .join(User, GardenInvitation.to_user_id == User.id)
        .filter(GardenInvitation.from_user_id == user.id, GardenInvitation.status == "pending")
        .order_by(GardenInvitation.created_at.desc())
        .all()
    )
    return [
        S.GardenInvitationView(
            id=inv.id, fromUserId=inv.from_user_id, toUserId=inv.to_user_id,
            fromName=user.name, fromEmail=user.email, toName=u.name, toEmail=u.email,
            createdAt=inv.created_at.isoformat(), direction="outgoing",
        )
        for inv, u in rows
    ]


@router.get("/invites")
def invites(db: DbSession = Depends(get_db), user: User = Depends(get_current_user)):
    return {"incoming": _incoming(db, user), "outgoing": _outgoing(db, user)}


@router.post("/invite", response_model=S.MsgOut)
async def invite(body: S.InviteRequest, db: DbSession = Depends(get_db), user: User = Depends(get_current_user)):
    email = body.email.strip().lower()
    if email == user.email.lower():
        raise HTTPException(400, "You are already growing in your own garden")

    if _guest_count(db, user.id) >= MAX_GUESTS:
        raise HTTPException(400, f"Garden is full ({GARDEN_CAPACITY} people including you)")

    target = db.query(User).filter(func.lower(User.email) == email).first()
    if not target:
        raise HTTPException(404, "No Bloom account with that email yet")

    already = db.query(GardenMember).filter(
        GardenMember.owner_id == user.id, GardenMember.member_id == target.id,
    ).first()
    if already:
        raise HTTPException(400, "They are already in your garden")

    existing = db.query(GardenInvitation).filter(
        GardenInvitation.from_user_id == user.id, GardenInvitation.to_user_id == target.id,
    ).first()
    if existing and existing.status == "pending":
        raise HTTPException(400, "Invitation already sent")

    reverse = db.query(GardenInvitation).filter(
        GardenInvitation.from_user_id == target.id,
        GardenInvitation.to_user_id == user.id,
        GardenInvitation.status == "pending",
    ).first()
    if reverse:
        raise HTTPException(400, "They already invited you — check your invitations")

    inv_id = existing.id if existing else uuid4().hex
    if existing:
        existing.status = "pending"
        existing.created_at = datetime.utcnow()
        existing.responded_at = None
    else:
        db.add(GardenInvitation(
            id=inv_id, from_user_id=user.id, to_user_id=target.id, status="pending",
        ))

    db.commit()
    await hub.publish(target.id, {
        "type": "invite-added",
        "data": {
            "id": inv_id, "fromUserId": user.id, "toUserId": target.id,
            "fromName": user.name, "fromEmail": user.email,
            "toName": target.name, "toEmail": target.email,
            "createdAt": datetime.utcnow().isoformat(), "direction": "incoming",
        },
    })
    return S.MsgOut()


@router.post("/invites/{invite_id}/accept", response_model=S.MsgOut)
async def accept(invite_id: str, db: DbSession = Depends(get_db), user: User = Depends(get_current_user)):
    inv = db.query(GardenInvitation).filter(
        GardenInvitation.id == invite_id,
        GardenInvitation.to_user_id == user.id,
        GardenInvitation.status == "pending",
    ).first()
    if not inv:
        raise HTTPException(404, "Invitation not found")

    from_id, to_id = inv.from_user_id, inv.to_user_id
    if _guest_count(db, from_id) >= MAX_GUESTS:
        raise HTTPException(400, "Their garden is full")
    if _guest_count(db, to_id) >= MAX_GUESTS:
        raise HTTPException(400, f"Your garden is full ({GARDEN_CAPACITY} people including you)")

    _ensure_membership(db, from_id, to_id)
    _ensure_membership(db, to_id, from_id)

    inv.status = "accepted"
    inv.responded_at = datetime.utcnow()

    db.query(GardenInvitation).filter(
        GardenInvitation.status == "pending",
        or_(
            and_(GardenInvitation.from_user_id == from_id, GardenInvitation.to_user_id == to_id),
            and_(GardenInvitation.from_user_id == to_id, GardenInvitation.to_user_id == from_id),
        ),
    ).delete(synchronize_session="fetch")

    db.commit()
    for uid in [from_id, to_id]:
        await hub.publish(uid, {"type": "invite-removed", "data": {"invitationId": invite_id}})
        await hub.publish(uid, {"type": "garden-changed", "data": {"reason": "invite-accepted"}})
    return S.MsgOut()


@router.post("/invites/{invite_id}/decline", response_model=S.MsgOut)
async def decline(invite_id: str, db: DbSession = Depends(get_db), user: User = Depends(get_current_user)):
    inv = db.query(GardenInvitation).filter(
        GardenInvitation.id == invite_id,
        GardenInvitation.to_user_id == user.id,
        GardenInvitation.status == "pending",
    ).first()
    if not inv:
        raise HTTPException(404, "Invitation not found")

    inv.status = "declined"
    inv.responded_at = datetime.utcnow()
    from_id = inv.from_user_id
    db.commit()

    for uid in [user.id, from_id]:
        await hub.publish(uid, {"type": "invite-removed", "data": {"invitationId": invite_id}})
    await hub.publish(from_id, {"type": "garden-changed", "data": {"reason": "invite-declined"}})
    return S.MsgOut()


@router.post("/invites/{invite_id}/cancel", response_model=S.MsgOut)
async def cancel(invite_id: str, db: DbSession = Depends(get_db), user: User = Depends(get_current_user)):
    inv = db.query(GardenInvitation).filter(
        GardenInvitation.id == invite_id,
        GardenInvitation.from_user_id == user.id,
        GardenInvitation.status == "pending",
    ).first()
    if not inv:
        raise HTTPException(404, "Invitation not found")

    to_id = inv.to_user_id
    db.delete(inv)
    db.commit()

    for uid in [user.id, to_id]:
        await hub.publish(uid, {"type": "invite-removed", "data": {"invitationId": invite_id}})
    return S.MsgOut()


@router.delete("/members/{member_id}", response_model=S.MsgOut)
async def remove_member(member_id: str, db: DbSession = Depends(get_db), user: User = Depends(get_current_user)):
    row = db.query(GardenMember).filter(
        GardenMember.id == member_id, GardenMember.owner_id == user.id,
    ).first()
    if not row:
        raise HTTPException(404, "Member not found")

    other_id = row.member_id
    db.query(GardenMember).filter(
        or_(
            and_(GardenMember.owner_id == user.id, GardenMember.member_id == other_id),
            and_(GardenMember.owner_id == other_id, GardenMember.member_id == user.id),
        ),
    ).delete(synchronize_session="fetch")
    db.commit()

    for uid in [user.id, other_id]:
        await hub.publish(uid, {"type": "garden-changed", "data": {"reason": "member-removed"}})
    return S.MsgOut()
